src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,194 @@
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Run;
+
+namespace Microsoft.DotNet.Tools;
+
+/// <summary>
+/// Used to build a virtual project file in memory to support <c>dotnet run file.cs</c>.
+/// </summary>
+internal sealed class VirtualProjectBuildingCommand
+{
+    public Dictionary<string, string> GlobalProperties { get; } = new(StringComparer.OrdinalIgnoreCase);
+    public required string EntryPointFileFullPath { get; init; }
+
+    public int Execute(string[] binaryLoggerArgs, LoggerVerbosity verbosity)
+    {
+        var binaryLoggers = GetBinaryLoggers(binaryLoggerArgs);
+        var consoleLogger = new ConsoleLogger(verbosity);
+        Dictionary<string, string?> savedEnvironmentVariables = new();
+        try
+        {
+            // Set environment variables.
+            foreach (var (key, value) in MSBuildForwardingAppWithoutLogging.GetMSBuildRequiredEnvironmentVariables())
+            {
+                savedEnvironmentVariables[key] = Environment.GetEnvironmentVariable(key);
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            // Setup MSBuild.
+            var projectCollection = new ProjectCollection(
+                GlobalProperties,
+                [.. binaryLoggers, consoleLogger],
+                ToolsetDefinitionLocations.Default);
+            var parameters = new BuildParameters(projectCollection)
+            {
+                Loggers = projectCollection.Loggers,
+                LogTaskInputs = binaryLoggers.Length != 0,
+            };
+            BuildManager.DefaultBuildManager.BeginBuild(parameters);
+
+            // Do a restore first (equivalent to MSBuild's "implicit restore", i.e., `/restore`).
+            // See https://github.com/dotnet/msbuild/blob/a1c2e7402ef0abe36bf493e395b04dd2cb1b3540/src/MSBuild/XMake.cs#L1838
+            // and https://github.com/dotnet/msbuild/issues/11519.
+            var restoreRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection, addGlobalProperties: static (globalProperties) =>
+                {
+                    globalProperties["MSBuildRestoreSessionId"] = Guid.NewGuid().ToString("D");
+                    globalProperties["MSBuildIsRestoring"] = bool.TrueString;
+                }),
+                targetsToBuild: ["Restore"],
+                hostServices: null,
+                BuildRequestDataFlags.ClearCachesAfterBuild | BuildRequestDataFlags.SkipNonexistentTargets | BuildRequestDataFlags.IgnoreMissingEmptyAndInvalidImports | BuildRequestDataFlags.FailOnUnresolvedSdk);
+            var restoreResult = BuildManager.DefaultBuildManager.BuildRequest(restoreRequest);
+            if (restoreResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            // Then do a build.
+            var buildRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection),
```

Comment:
Why are we creating two separate project instances?

Reply:
Because we need to set additional global properties. As far as I understand MSBuild APIs, this is the only way to do it.

Reply:
BuildRequestData has an overload with global properties

Reply:
> BuildRequestData has an overload with global properties

Those ones only take physical file paths though, right? (And they read the XML from the path passed in.) I need to use a BuildRequestData overload that takes the XML data because my project file doesn't exist on disk.

Reply:
Hmm...I think you're right. You could consider adding a new API for that to MSBuild, since that seems like it would be very easy to implement, but I'd consider that a v3 change not necessary for this PR.